src/tables.jl

```diff
@@ -1,189 +1,424 @@
 Tables.istable(::Type{<:File}) = true
-Tables.schema(f::File)  = Tables.Schema(f.names, f.types)
-
-Tables.columnaccess(::Type{File{t, c, I, P, KW}}) where {t, c, I, P, KW} = c
-
-struct Columns
-    names
-    columns
-end
-Base.propertynames(c::Columns) = getfield(c, 1)
-function Base.getproperty(c::Columns, nm::Symbol)
-    getfield(c, 2)[findfirst(x->x==nm, getfield(c, 1))]
-end
-
-function Tables.columns(f::File{transpose, true}) where {transpose}
-    len = length(f)
-    types = f.types
-    columns = Any[Tables.allocatecolumn(T, len) for T in types]
-    funcs = [get(FUNCTIONMAP, parsingtype(T), FUNC_ANY[]) for T in types]
-    transpose && (f.positions .= f.originalpositions)
-    for row = 1:len
-        @inbounds Parsers.fastseek!(f.io, f.positions[row])
-        f.currentrow[] = row
-        f.lastparsedcol[] = 0
-        f.lastparsedcode[] = Parsers.SUCCESS
-        for col = 1:length(types)
-            @inbounds ccall(funcs[col], Cvoid, (Any, Any, Cssize_t, Cssize_t), columns[col], f, col, row)
+Tables.columnaccess(::Type{<:File}) = true
+Tables.schema(f::File)  = nothing
+
+# a mapping of TypeCode to its corresponding values array in Array struct
+# enables doing things like getfield(A::CSV.Array, TYPEINDEX[col.typecode])
+const TYPEINDEX = Dict(
+    EMPTY    => 0,
+    MISSING  => 6,
+    INT      => 7,
+    FLOAT    => 8,
+    DATE     => 9,
+    DATETIME => 10,
+    BOOL     => 11,
+    (INT | MISSING)      => 12,
+    (FLOAT | MISSING)    => 13,
+    (DATE | MISSING)     => 14,
+    (DATETIME | MISSING) => 15,
+    (BOOL | MISSING)     => 16,
+)
+
+mutable struct Array
+    offsets::Vector{UInt64}
+    lengths::Vector{UInt32}
+    typecode::TypeCode
+    escapestrings::Bool
+    lastref::UInt32
+
+    missings::Vector{Missing}
+    ints::Vector{Int64}
+    floats::Vector{Float64}
+    dates::Vector{Date}
+    datetimes::Vector{DateTime}
+    bools::Vector{Bool}
+    
+    intsm::Vector{Union{Int64, Missing}}
+    floatsm::Vector{Union{Float64, Missing}}
+    datesm::Vector{Union{Date, Missing}}
+    datetimesm::Vector{Union{DateTime, Missing}}
+    boolsm::Vector{Union{Bool, Missing}}
+
+    refs::Vector{UInt32}
+    pool::Dict{String, UInt32}
+    poolm::Dict{Union{String, Missing}, UInt32}
+
+    Array(len) = new(Vector{UInt64}(undef, len), Vector{UInt32}(undef, len), EMPTY, false, 0)
+end
+
+# "materialize" a CSV.Array as a real Array/CategoricalArray/PooledArray/StringArray
+# and resize the resulting array according to the # of rows we actually parsed
+function vec(buf, x::Array, rows, E, pool, categorical)
```

Comment:
It's quite confusing to have a type named `Array` since you can't tell which type this refers to (same below). Maybe use a different name?